### ISSUE TITLE
Encoding Problem

### ISSUE THREAD

**Issue by farid220** [synfron/ReshaperForBurp#15]
Hi 
First of all I wanna thank you for this great extension I was trying to implement sth like this then I hit this BApp of yours.
I'm having trouble mocking a http response which contains ISO-8859-1 characters and when I just copy and paste my desired body and check the response I see weird stuff (encoding wired chars) exactly where those characters exists. I'd be grateful if you could consider adding support of multiple encodings in your application because technically not all websites are English. Strangely enough I couldn't even reverse the encoded message to its original format (It might be doubly encoded or sth)  

I might know how to solve my own specific problem though I think adding a .encode(iso-8859-1) before StringToByte method (in converting string to byte for assembling ResponseMessage to send) might do it however, I don't know how to compile your source code to get the jar file.

If you need I can provide ambiguous encoded message compared to its original format for further analysis.

**Comment by ddwightx** [comment]
I'm glad the extension has been helpful to you.

Do you have an example of an Burp extension (https://github.com/PortSwigger) that supports this properly in your experience?

Can you give an example of text that would exhibit the issue and reproduction steps that will show the issue in Reshaper when this text is used? Should be able to attach a file containing the text here.

FYI, I have added development instructions to the [Readme](https://github.com/synfron/ReshaperForBurp#development).

**Comment by farid220** [comment]
Hi thanks for your reply
Honestly, I haven't tried many BApps and I don't know any other specifically which supports multiple encoding.

Sure, I'm uploading 2 text files which contains corrupted body and original one and a picture of RULES tab.
I wonder why I couldn't find building instruction LOL 
I'm trying to apply the solution I mentioned in previous post, and I will keep you informed about the outcome.
Basically as an example the word "حداکثر توان" which is Persian gets converted to "-/'©+1 *H'F"

TrueBody:
[https://0bin.net/paste/aFYL7umh#x43B5kYTX1NeQ9bSscm9de+TMNaNb-0MWtHOJ6FqXFA](url)

Problematic:
[https://0bin.net/paste/etS6SItP#B0a02dge4nvuSvd08GblH6ouT2B8i1dCvFj7+50wxVX](url)

Rules Tab Screenshot:
[https://postimg.cc/rKwFQ8JT](url)

**Comment by ddwightx** [comment]
iso-8859-1 doesn't appear to be the right encoding. It doesn't have codes to represent the characters you provided in your example (https://www.charset.org/charsets/iso-8859-1).

Not sure how it works in Python, but in Java, all Strings are internally represented as Unicode, so there isn't a need to transition between encodings while text is still in String form (unlike `respS = respS .encode('iso-8859-1')` which appears to be changing the internal string encoding since the output is still a string). In Java, the encoding should only come into play when converting the String to bytes or from bytes to a String. In our case, Reshaper is using the recommended `stringToBytes` method provided by Burp Suite rather than using Java's built-in library for encoding. It appears that the problem is in the way their method is handling the String. It does not occur using Java's built-in library. I will have to see what Burp Suite folks say about this issue.

**Comment by farid220** [comment]
Hi thanks for digging into this problem
Yes there are some ambiguity in charset of encoded strings. (I think it is doubly encoded or sth) or it could be stringToBytes method misinterpreting the encoding of the data. 
About StringToSting conversion you are right. The data retrived from burpAPI methods, is first decoded from byte to string and then they are reassembled again into byte and sent back to burpAPI, but in non-ASCII it's somehow needed that the same STRING gets encoded again and the outcome needs the stringtobyte . Basically one extra step is required (a encoding conversion of let's say StringToString). I was trying to implement the equivalent code of Python in Java and I got "Unhandled exception type UnsupportedEncodingException" . I fixed it using throws but I realized I have to fix every single code that uses setbody method which kinda feels wrong 

ReshaperForBurp-release\src\main\java\synfron\reshaper\burp\core\messages\entities\HttpResponseMessage.java I believe is the actual script we need to apply our changes and I change setBody as following:

`
import java.io.UnsupportedEncodingException;



public void setBody(String body) throws UnsupportedEncodingException {
        String decodedToUTF8 = new String(body.getBytes("ISO-8859-1"), "UTF-8");
        this.body = new HttpBody(TextUtils.stringToBytes(decodedToUTF8));
        //this.body = new HttpBody(TextUtils.stringToBytes(body));
        changed = true;
    }`
When I try to compile, it fails saying "UnsupportedEncodingException" where ever setBody is used so I'm trying to feel that this is not the way of StringToString conversion. So I would appreciate it if you could help me to implement the the equivalent of respS = respS .encode('iso-8859-1') in Java. It seems much easier in Python but seemingly in Java matter of encoding  applies where we have Bytes and String form has its challenges.

**Comment by ddwightx** [comment]
I believe you miss interpreted my comment above. First, there is no need to worry about String to String encoding. Second, `ISO-8859-1` would be the incorrect encoding to use as an output because it does not support all of the characters in the example text you gave.

Below is the git patch to change what Reshaper uses for String to Byte conversion to use UTF-8.
```
diff --git a/src/main/java/synfron/reshaper/burp/core/utils/TextUtils.java b/src/main/java/synfron/reshaper/burp/core/utils/TextUtils.java
index 37083bf..32ec8e5 100644
--- a/src/main/java/synfron/reshaper/burp/core/utils/TextUtils.java
+++ b/src/main/java/synfron/reshaper/burp/core/utils/TextUtils.java
@@ -109,7 +109,7 @@ public class TextUtils {
     }
 
     public static byte[] stringToBytes(String value) {
-        return BurpExtender.getCallbacks().getHelpers().stringToBytes(StringUtils.defaultString(value));
+        return StringUtils.defaultString(value).getBytes(StandardCharsets.UTF_8);
     }
 
     public static String bufferAppend(String text1, String text2, String suffix, int maxLength) {

```

**Comment by farid220** [comment]
Hi Thanks for your help I managed to compile it using Gradle 6.8 and Java 15 just as you mentioned. However compiling using the latest versions of Java and Gradle failed saying "No duplicate handling strategy has been set" and I know it could be fixed by configuring some settings but I just wanted to avoid complexities.

 The committed cahnge works just fine and I haven't seen any problems so I'd recommend considering this Method over the previous one because it just supports everything I suppose, and as you are the developer still you can decide better.

Thank you so much you for this great help and amazing extension you developed. It is really interactive and modular.

**Comment by ddwightx** [comment]
Glad the change worked for you. I have left a message on the Burp Suite forums concerning the issue you found. Hopefully they will fix that issue so this change doesn't become necessary.

**Comment by ddwightx** [comment]
Please see release v1.6.0 (Not yet available in the BApp Store). An option to change the default encoding has been added to the Settings. A new Then action (Set Encoding) was also added for more targeted use cases.

https://github.com/synfron/ReshaperForBurp/releases/tag/v1.6.0

Thanks for reporting this issue. 🙂

**Comment by farid220** [comment]
Hi
Thank you so much for fixing this issue and helping me to get what I needed.
I just downloaded this new release and now I see the encoding option. But unfortunately I can't get it working and I still get the previous problematic output. I believe I couldn't configure it properly so I would be grateful if you could tell me how to configure it to make it work in my specific case.

**Comment by ddwightx** [comment]
Looks like there is a bug in the request encoding detection (is in the wrong location).

As a workaround for now, to emulate the same general functionality create a new Rule like below at the beginning of your Rules list:
When - Event Direction - Request
When - Content Type - JSON, XML, URL Encoded
Then - Set Encoding - Auto-detect

FYI, There are some texts where auto-detect will not work well because auto-detection is requires heuristics. To get the most accurate output, I'd recommend only setting the encoding in locations to handle very specific situations (as is possible with the Then), and using a specific encoding that is known to have the expected output.

**Comment by ddwightx** [comment]
@farid220 Can you please try https://github.com/synfron/ReshaperForBurp/releases/tag/v1.6.0-next and see if it fixes your issue?

**Comment by farid220** [comment]
I'm so sorry I don't know why Github didn't send me notification about your message. I tested version 1.6.0 and it works just fine (except the very default Encoding setting).
I see you committed 20 hours ago and I will test this one. 

I encountered sth else though and I think it might be my inability to set it up.
The thing is there is a URL of an image which basically returns the image (PNG).
Burp's interceptor shows the header with some weird characters (Which is reasonable because it's image in byte format)  

Now I want to implement it in Reshaper but I just can't get it done. I copy pasted the exact data from Burp interceptor but it just didn't work. 

Could you please elaborate on how can I implement this in Reshaper?

**Comment by ddwightx** [comment]
v1.6.1 is release to the BApp Store to fix the encoding issue. Please open a discussion or new issue for your image problem.
